File: polyglot/runtime_template_resolver/py/src/runtime_template_resolver/validator.py
import re
from .errors import ValidationError

def validate_placeholder(placeholder: str) -> None:
    if not placeholder or not placeholder.strip():
        raise ValidationError("Placeholder cannot be empty")
    
    # Basic validation
    if not re.match(r'^[\w\-\.\[\]"\']+$', placeholder):
        raise ValidationError(f"Invalid characters in placeholder: {placeholder}")

    if ".." in placeholder:
        raise ValidationError(f"Invalid path (empty segment): {placeholder}")

File: polyglot/runtime_template_resolver/mjs/src/validator.ts
import { ValidationError } from './errors.js';

export function validatePlaceholder(placeholder: string): void {
    if (!placeholder || placeholder.trim() === '') {
        throw new ValidationError('Placeholder cannot be empty');
    }
    // Basic validation: must be valid path
    // Allowing dots, brackets, quotes, alphanumeric, underscore, hyphen
    if (!/^[\w\-\.\[\]"']+$/.test(placeholder)) {
        throw new ValidationError(`Invalid characters in placeholder: ${placeholder}`);
    }

    if (placeholder.includes('..')) {
        throw new ValidationError(`Invalid path (empty segment): ${placeholder}`);
    }
}
